### ISSUE TITLE
Remove fractal-common and remove/review API-call validation

### ISSUE THREAD

**Issue by tcompa** [fractal-analytics-platform/fractal#548]
# TLDR

Sorry for the verbose issue; the actionable part is quite simple, but I wanted to make sure we are aware of all options.

1. We (with @mfranzon) suggest to remove the existing client-side validation of API request/response bodies, and then remove the dependency on fractal-common git submodule.
2. If we decide that redundant validation is actually required, I suggest we explore different options that rely on standard protocols (i.e. OpenAPI), rather than on a git submodule (related to #210).

Both options would make #545 (i.e. going from three repositories to a single one) irrelevant.

# Current status

1. Internal client logic for preparing the request body
2. Client-side request-body validation
3. API request
4. Server-side request-body validation (failure results in a 422 response error, with details)
5. Internal endpoint logic
6. Server-side response-body validation (failure results in a 500 internal error -> bug in fractal-server)
7. API response
8. Client-side response-body validation
9. Response is processed and shown to the user

Steps 2 is partially redundant and step 8 is fully redundant, as they replicate the validation that already takes place server-side. This duplication also requires code duplication for the serialization models, which we achieve through the `fractal-common` git submodule. This shared repository **is not versioned**, and then each change must be immediately propagated to both server and client - which is extremely time-consuming during developement (a tiny addition results in three PRs, to be created and merged only when the corresponding CI actions are complete). The alternative option (versioning of fractal-common) is fully out-of-scope, as it would increase the package complexity even more.

# Main proposal

We propose to fully remove steps 2 and 8 above, and to get rid of fractal-common.

## Improvements

* We can fully transform `fractal-common` into a simple subfolder in `fractal-server`.
* We have a single source of truth (it's not possible to have different `fractal-common` versions in server and client).
* Client-side response-body validation is really fully redundant and it should simply be removed. The reason is that the client cannot include any additional validation rule on top of the ones that are implemented in fractal-server.
* We greatly scale down the dependency on Pydantic in fractal-client, which will make the future transition to V2 easier; and, most important, it will be possible to switch to V2 even before we do it for `fractal-server` (there are caveats, but they are out-of-scope here).

## Possible drawbacks

* There may be situations where screening API calls client-side is useful. Say that a user wants to add 1000 resources to a dataset, and puts the `fractal dataset add-resource` command in a `for` loop. Say that the resource paths are all invalid (because they are not absolute). This will result in 1000 API calls, which could be prevented if we had kept client-side validation in-place.
Note that this issue is easily mitigated: there are not many commands that can lead to an invalid request-body, and we could look at them one by one and add additional logic in the client internal functions if appropriate. 

## To-do list

* Fully remove validation of request bodies
* Fully remove validation of request bodies
* Remove any reference to fractal-common
* Go through client-made API calls, look for possible validation errors, and decide whether we can add some additional checks (e.g. resource paths must be absolute).
* Improve displaying of server-side validation errors (it is already there, but we could make it more clear).

# Alternative proposal

If we decide otherwise, and we want to keep client-side validation, we should still somehow improve the current status and move away from `fractal-common`. This alternative proposal is just a stub to mention how things could be done differently, but we can go into details if needed.

`fractal-server` already has a standardized way to communicate the valid request bodies, through the OpenAPI specs. If we need to validate a request, that must go through this protocol.
Note the connection with #210: if we want to make any client-side validation, we must use the API specs the server is using; any API specs which is defined client-side cannot be trusted.

## Example
Just as an example, if we want to stick with Pydantic, a command like
```
datamodel-codegen --url http://localhost:8000/openapi.json --output model.py --input-file-type openapi
```
automatically generates (via https://koxudaxi.github.io/datamodel-code-generator) a file `model.py` which includes all the serialization Pydantic models (`TaskCrete`, `TaskUpdate`, ...).

## Other options

I'm sure there exist multiple other options, e.g. validating payloads directly against OpenAPI JSON Schemas, without the additional step of converting them to Pydantic models.

**Comment by jluethi** [comment]
Approved

**Comment by tcompa** [comment]
To do:
- [ ] Fully remove validation of request bodies
- [ ] Fully remove validation of response bodies (aka `coerce` argument)
- [ ] Remove all imports from `common`
- [ ] Update a lot of likely failing tests
- [ ] Review `type` attributes of argparse parsers, and include as many as possible

---

Lower priority:

- [ ] Go through client-made API calls, look for possible validation errors, and decide whether we can add some additional checks (e.g. resource paths must be absolute).
- [ ] Improve displaying of server-side validation errors (it is already there, but we could make it more clear - ref https://github.com/fractal-analytics-platform/fractal/blob/main/fractal/response.py).
- [ ] Remove pydantic from fractal-client

**Comment by tcompa** [comment]
> Go through client-made API calls, look for possible validation errors, and decide whether we can add some additional checks (e.g. resource paths must be absolute).

After some discussion with @ychiucco, we are taking this back.

The pro of this action would be to cover a very small part of all the possible server errors (essentially only some empty strings or non-absolute paths), with the negative consequence that we would be once again duplicating some logic (i.e. the definition of what API calls are valid) and then we'd incur the risk of out-of-sync server/client.

> Improve displaying of server-side validation errors (it is already there, but we could make it more clear - ref https://github.com/fractal-analytics-platform/fractal/blob/main/fractal/response.py).

As of https://github.com/fractal-analytics-platform/fractal/pull/551/commits/7c2c66dfa5947b46a319940619b51b80ed16b113, request-body validation errors are handled in a custom way, and the fractal-client log looks like
![Screenshot from 2023-09-25 11-59-39](https://github.com/fractal-analytics-platform/fractal/assets/3862206/45b0af00-40e2-4be6-9734-8884efac5920)
